sql-plugin/src/main/scala/org/apache/spark/sql/rapids/zorder/ZOrderRules.scala

```diff
@@ -0,0 +1,113 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.rapids.zorder
+
+import com.nvidia.spark.rapids.{ExprChecks, ExprMeta, ExprRule, GpuExpression, GpuRangePartitioner, GpuSorter, RepeatingParamCheck, ShimLoader, TypeSig, UnaryExprMeta}
+import com.nvidia.spark.rapids.GpuOverrides.{expr, pluginSupportedOrderableSig}
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{AttributeReference, Expression, SortOrder, UnaryExpression}
+import org.apache.spark.sql.catalyst.expressions.codegen.LazilyGeneratedOrdering
+
+/**
+ * These are rules to be able to replace zorder operators. For now they all come from deltalake,
+ * but not all versions of deltalake support zordering so the simplest solution is to pick them
+ * apart with reflection.  That way if they exist on the classpath we can replace them.
+ */
+object ZOrderRules {
+  def exprs: Map[Class[_ <: Expression], ExprRule[_ <: Expression]] = {
+    try {
+      val interleaveClazz =
+        ShimLoader.loadClass("org.apache.spark.sql.delta.expressions.InterleaveBits")
+            .asInstanceOf[Class[Expression]]
+      val interleaveRule = expr[Expression](
+        "Interleave bit as a part of deltalake zorder",
+        ExprChecks.projectOnly(
+          TypeSig.BINARY,
+          TypeSig.BINARY,
+          repeatingParamCheck =
+            Some(RepeatingParamCheck("input",
+              TypeSig.INT,
+              TypeSig.INT))),
+        (a, conf, p, r) => new ExprMeta[Expression](a, conf, p, r) {
+          override def convertToGpu(): GpuExpression =
+            GpuInterleaveBits(childExprs.map(_.convertToGpu()))
+        })
+
+      val partExprClass =
+        ShimLoader.loadClass("org.apache.spark.sql.delta.expressions.PartitionerExpr")
+            .asInstanceOf[Class[Expression]]
+      val partExprRule = expr[UnaryExpression](
+        "Partitioning on a single column for deltalake zorder",
+        ExprChecks.unaryProject(
+          TypeSig.INT,
+          TypeSig.INT,
+          // Same as RangePartitioningExec
+          (pluginSupportedOrderableSig + TypeSig.DECIMAL_128 + TypeSig.STRUCT).nested(),
+          TypeSig.orderable),
+        (a, conf, p, r) => new UnaryExprMeta[UnaryExpression](a, conf, p, r) {
+          def getBoundsNOrder: (Array[InternalRow], SortOrder) = {
+            // We have to pull pick apart the partitioner apart, but we have to use reflection
+            // to do it, because what we want is private
+            val parterField = partExprClass.getDeclaredField("partitioner")
+            parterField.setAccessible(true)
+            val parter = parterField.get(a)
+            val partClass = parter.getClass
```

Comment:
let us use https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/reflect/FieldUtils.html#readField-java.lang.reflect.Field-java.lang.Object-boolean- for this boilerplate code

Reply:
I am not so sure that I do what to use this library. Reading through the libraries code it is explicitly catching some exceptions and returning nulls. This makes the error handling and error detection more difficult. 

I will need to expand the code to catch errors to include `| _: IllegalArgumentException | _: NullPointerException` not too big of a deal, but it might mask other errors that I want to see/understand. and/or I need to add in null checks after each get `getDeclaredField` operation or it's equivalent. I don't think it will reduce the amount of code needed and it is not that much boiler plate code anyways.